File: LLM_Translation/C_programs/delete.c
#include<stdio.h>
#include<stdlib.h>

struct Node
{
    int key;
    struct Node *left;
    struct Node *right;
    int height;
};

int getBalance(struct Node *N)
{
    if (N == NULL)
        return 0;
    return height(N->left) - height(N->right);
}

struct Node *leftRotate(struct Node *x)
{
    struct Node *y = x->right;
    struct Node *T2 = y->left;
 
    // Perform rotation and Update Heights
    x->right = T2;
    x->height = max(height(x->left), height(x->right))+1;
    y->left = x;
    y->height = max(height(y->left), height(y->right))+1;
 
    // Return new root
    return y;
}

struct Node *rightRotate(struct Node *y)
{
    struct Node *x = y->left;
    struct Node *T2 = x->right;
 
    // Perform rotation and Update Heights
    y->left = T2;
    y->height = max(height(y->left), height(y->right))+1;
    x->right = y;
    x->height = max(height(x->left), height(x->right))+1;
 
    // Return new root
    return x;
}
 

struct Node* deleteNode(struct Node* root, int key)
{
    // STEP 1: PERFORM STANDARD BST DELETE
 
    if (root == NULL)
        return root;
 
    // If the key to be deleted is smaller than the
    // root's key, then it lies in left subtree
    if ( key < root->key )
        root->left = deleteNode(root->left, key);
 
    // If the key to be deleted is greater than the
    // root's key, then it lies in right subtree
    else if( key > root->key )
        root->right = deleteNode(root->right, key);
 
    // if key is same as root's key, then This is
    // the node to be deleted
    else
    {
        // node with only one child or no child
        if( (root->left == NULL) || (root->right == NULL) )
        {
            struct Node *temp = root->left ? root->left :
                                             root->right;
 
            // No child case
            if (temp == NULL)
            {
                temp = root;
                root = NULL;
            }
            else // One child case
             *root = *temp; // Copy the contents of
                            // the non-empty child
            free(temp);
        }
        else
        {
            // node with two children: Get the inorder
            // successor (smallest in the right subtree)
            struct Node* temp = minValueNode(root->right);
 
            // Copy the inorder successor's data to this node
            root->key = temp->key;
 
            // Delete the inorder successor
            root->right = deleteNode(root->right, temp->key);
        }
    }
 
    // If the tree had only one node then return
    if (root == NULL)
      return root;
 
    // STEP 2: UPDATE HEIGHT OF THE CURRENT NODE
    root->height = 1 + max(height(root->left),
                           height(root->right));
 
    // STEP 3: GET THE BALANCE FACTOR OF THIS NODE (to
    // check whether this node became unbalanced)
    int balance = getBalance(root);
 
    // If this node becomes unbalanced, then there are 4 cases
 
    // Left Left Case
    if (balance > 1 && getBalance(root->left) >= 0)
        return rightRotate(root);
 
    // Left Right Case
    if (balance > 1 && getBalance(root->left) < 0)
    {
        root->left =  leftRotate(root->left);
        return rightRotate(root);
    }
 
    // Right Right Case
    if (balance < -1 && getBalance(root->right) <= 0)
        return leftRotate(root);
 
    // Right Left Case
    if (balance < -1 && getBalance(root->right) > 0)
    {
        root->right = rightRotate(root->right);
        return leftRotate(root);
    }
 
    return root;
}

int main(){}

File: LLM_Translation/Rust_programs_not_compiled/delete.rs
use std::cmp;

struct Node {
    key: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
    height: i32,
}

fn height(node: &Option<Box<Node>>) -> i32 {
    match node {
        Some(n) => n.height,
        None => 0,
    }
}

fn get_balance(n: &Option<Box<Node>>) -> i32 {
    match n {
        Some(node) => height(&node.left) - height(&node.right),
        None => 0,
    }
}

fn left_rotate(x: Option<Box<Node>>) -> Option<Box<Node>> {
    if let Some(mut x_node) = x {
        if let Some(mut y_node) = x_node.right.take() {
            let t2 = y_node.left.take();
            x_node.right = t2;
            x_node.height = cmp::max(height(&x_node.left), height(&x_node.right)) + 1;
            y_node.left = Some(x_node);
            y_node.height = cmp::max(height(&y_node.left), height(&y_node.right)) + 1;
            return Some(y_node);
        }
    }
    None
}

fn right_rotate(y: Option<Box<Node>>) -> Option<Box<Node>> {
    if let Some(mut y_node) = y {
        if let Some(mut x_node) = y_node.left.take() {
            let t2 = x_node.right.take();
            y_node.left = t2;
            y_node.height = cmp::max(height(&y_node.left), height(&y_node.right)) + 1;
            x_node.right = Some(y_node);
            x_node.height = cmp::max(height(&x_node.left), height(&x_node.right)) + 1;
            return Some(x_node);
        }
    }
    None
}

fn min_value_node(node: &Option<Box<Node>>) -> Option<Box<Node>> {
    let mut current = node.clone();
    while let Some(ref current_node) = current {
        if current_node.left.is_none() {
            break;
        }
        current = current_node.left.clone();
    }
    current
}

fn delete_node(root: Option<Box<Node>>, key: i32) -> Option<Box<Node>> {
    let mut root = root;
    if root.is_none() {
        return root;
    }
    if let Some(root_ref) = root.as_mut() {
        if key < root_ref.key {
            root_ref.left = delete_node(root_ref.left.take(), key);
        } else if key > root_ref.key {
            root_ref.right = delete_node(root_ref.right.take(), key);
        } else {
            if root_ref.left.is_none() || root_ref.right.is_none() {
                let temp = if root_ref.left.is_some() {
                    root_ref.left.take()
                } else {
                    root_ref.right.take()
                };
                if temp.is_none() {
                    root = None;
                } else {
                    *root_ref = *temp.unwrap();
                }
            } else {
                let temp = min_value_node(&root_ref.right).unwrap();
                root_ref.key = temp.key;
                root_ref.right = delete_node(root_ref.right.take(), temp.key);
            }
        }
    }
    if root.is_some() {
        if let Some(root_ref) = root.as_mut() {
            root_ref.height = cmp::max(height(&root_ref.left), height(&root_ref.right)) + 1;
            let balance = get_balance(root.as_ref());
            if balance > 1 && get_balance(root_ref.left.as_ref()) >= 0 {
                return right_rotate(root);
            }
            if balance > 1 && get_balance(root_ref.left.as_ref()) < 0 {
                root_ref.left = left_rotate(root_ref.left.take());
                return right_rotate(root);
            }
            if balance < -1 && get_balance(root_ref.right.as_ref()) <= 0 {
                return left_rotate(root);
            }
            if balance < -1 && get_balance(root_ref.right.as_ref()) > 0 {
                root_ref.right = right_rotate(root_ref.right.take());
                return left_rotate(root);
            }
        }
    }
    root
}

fn main() {}